### PR TITLE
Schedule same transforms only once per artifact

### PR DESCRIPTION
**Pull request by wolfs**

### CODE REVIEW

**Comment by oehme** [comment]
As far as I can see this still means all the transforms are printed on every build, even though nothing is being transformed. I think the build operation should be [down here](https://github.com/gradle/gradle/blob/50bc98e7bb6017b87bdfcce1d82acbcc89cc7d74/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformedFileCache.java#L160) when we actually do something. Getting it from memory should not result in logging on my opinion.

**Comment by wolfs** [comment]
@oehme We want to get closer to tasks with artifact transforms. That means for logging that artifact transforms would be logged similar to tasks, too. For the plain console we would show the each artifact transform and the state, for example if it was from the cache or not, similarly to task headers. We only would see them in the status line for the rich console. Does this make sense to you?

This PR does not address the logging at all, it just makes sure we schedule less transforms as part of the task graph.

**Comment by oehme** [comment]
Well this PR addresses the logging by not doing the same transform again and again, or did I misinterpret that?

I see where you're going with that, it just feels a bit verbose. There are way more transformed artifacts than tasks. We already had users complaining about artifact downloads in the plain console. Showing transforms in persistent log output (even though they didn't execute anything) may lead to similar complaints.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java [hunk 1]
```diff
@@ -51,23 +50,12 @@
...
-                current = new InitialTransformInfo(transformer, artifact);
-            } else {
-                current = new ChainedTransformInfo(transformer, previous);
-            }
-            previous = current;
-        } while (iterator.hasNext());
-        return current;
+    public static TransformInfo chained(List<UserCodeBackedTransformer> transformerChain, TransformInfo previous) {
+        return new ChainedTransformInfo(transformerChain.get(transformerChain.size() - 1), previous);
+    }
```
Comment (lptr) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java]: This should not receive a `List` only to get the last element of it...

**Comment by lptr** [comment]
@oehme we are moving one step at a time. The plan is to rework the transform chains (we currently don't have a nice way to display those either) and that will allow us to print logs for transforms better. But for 4.10 this PR is what we can comfortably do.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java (hunk 1 above)
Comment (wolfs) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java]: I am now passing the last element of the list.

**Pull request merged**